code/game/machinery/mac_cannon.dm

```diff
@@ -0,0 +1,210 @@
+	if(breech.loaded_shell)
+		if(breech.loaded_objects.len > 1) // if there is a shell and other shit in the barrel, blow it up
+			explosion(breech,1,2,6)
+		else
+			var/obj/item/projectile/ship_projectile/mac_round/M = PoolOrNew(breech.loaded_shell.projectile,get_step(src,dir))
+			M.set_data(breech.loaded_shell.damage,1,0,target,breech.loaded_shell.armed)
+			M.setDir(src,dir)
+			M.starting = src.loc
+			M.fire()
+			fire_sound()
+		var/obj/item/weapon/twohanded/required/shell_casing/C = new breech.loaded_shell.casing
+		C.forceMove(breech)
+		qdel(breech.loaded_shell)
+		breech.loaded_objects -= breech.loaded_shell
+		breech.loaded_shell = null
+
+	else //otherwise shoot whatever is in the barrel
+		for(var/atom/movable/A in breech.loaded_objects)
+			A.forceMove(get_step(src,dir))
+			if(istype(A,/mob/living/))
+				var/mob/living/M = A
+				M.Weaken(5)
+			var/atom/throw_at = get_edge_target_turf(src, dir)
+			A.throw_at_fast(throw_at, 500, 1)
+
+			sleep(2)
+
+
+/obj/machinery/mac_barrel/proc/toggle_hatch() //just moves the functionality of the massdriver control into the mac cannon, cleaned up a bit
+	var/list/activated_doors = list()
+	for(var/obj/machinery/door/poddoor/M in machines)
+		if(M.id == id)
+			activated_doors += M
+			spawn(0) M.open()
+	sleep(50)
+
+	for(var/obj/machinery/door/poddoor/M in activated_doors)
+		spawn(0) M.close()
+
+
+/obj/machinery/mac_barrel/proc/fire_sound()
+	playsound(loc, 'sound/weapons/flashbang.ogg', 100, 1)
+	for(var/mob/living/carbon/human/H in range(5,src))
+		if(H.check_ear_prot()) continue
+		if(!istype(H.loc.loc,/area/shuttle/ftl/munitions)) continue
+
+		H.show_message("<span class='warning'>BANG</span>", 2)
+		H.Weaken(5)
+		H << sound('sound/weapons/flash_ring.ogg',0,1,0,100) //copied from flashbangs
+		H.setEarDamage(H.ear_damage + rand(0, 5), max(H.ear_deaf,15))
+		if (H.ear_damage >= 15)
+			H << "<span class='warning'>Your ears start to ring badly!</span>"
+			if(prob(H.ear_damage - 10 + 5))
+				H << "<span class='warning'>You can't hear anything!</span>"
+				H.disabilities |= DEAF
+		else
+			if (H.ear_damage >= 5)
+				H << "<span class='warning'>Your ears start to ring!</span>"
+
+/obj/machinery/mac_barrel/proc/find_breech()
+	for(var/obj/machinery/mac_breech/P in get_step(src,turn(dir,180)))
+		if(istype(P))
+			breech = P
+			return
+
+
+/obj/machinery/mac_breech
+	name = "\improper MAC cannon breech"
+	desc = "You should probably not put your hands in this thing during use."
+	icon = 'icons/obj/stationobjs.dmi'
+	icon_state = "mac_breech"
+
+	var/obj/structure/loader/loader = null
+	var/list/loaded_objects = list()
+	var/obj/structure/shell/loaded_shell = null
+
+	var/charge_process = 100
+
+
+	density = 1
+	anchored = 1
+
+/obj/machinery/mac_breech/New()
+	..()
+	var/obj/item/weapon/circuitboard/machine/B = new /obj/item/weapon/circuitboard/machine/mac_cannon(null)
+	B.apply_default_parts(src)
+	RefreshParts()
+
+
+/obj/item/weapon/circuitboard/machine/mac_cannon
```

Comment:
Isn't there a separate file for defining machine circuitboards or am I thinking of computers?